RecoMTD/MeasurementDet/src/MTDDetLayerMeasurements.cc

```diff
@@ -0,0 +1,397 @@
+	for (ME0SegmentCollection::const_iterator rechit = range.first; 
+	     rechit!=range.second; ++rechit){
+	  LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "On ME0 iteration " <<std::endl;
+	  result.push_back(MuonTransientTrackingRecHit::specificBuild(geomDet,&*rechit));
+	}
+	LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "Number of ME0 rechits = " << result.size()<<std::endl;
+      }
+  }
+*/
+/*
+// FIXME, these might be needed for debugging
+  else {
+    // wrong type
+    throw cms::Exception("MTDDetLayerMeasurements") << "The DetLayer with det " << geoId.det() << " subdet " << geoId.subdetId() << " is not a valid MTD DetLayer. ";
+    LogDebug("Track|RecoMTD|MTDDetLayerMeasurements")<< "The DetLayer with det " << geoId.det() << " subdet " << geoId.subdetId() << " is not a valid MTD DetLayer. ";
+  }
+*/
+  /*if (enableME0Measurement){
+    LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "(ME0): enabled"<<std::endl;
+  }
+
+  if (enableGEMMeasurement){
+    LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "(GEM): enabled"<<std::endl;
+    }*/
+  return result;
+}
+
+void MTDDetLayerMeasurements::checkMTDRecHits()
+{
+  LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "Checking MTD RecHits";
+  checkEvent();
+  auto cacheID = theEvent->cacheIdentifier();
+  if (cacheID == theMTDEventCacheID) return;
+
+  {
+    theEvent->getByToken(mtdToken_, theMTDRecHits);
+    theMTDEventCacheID = cacheID;
+  }
+  if(!theMTDRecHits.isValid())
+  {
+    throw cms::Exception("MTDDetLayerMeasurements") << "Cannot get MTD RecHits";
+    LogDebug("Track|RecoMTD|MTDDetLayerMeasurements") << "Cannot get MTD RecHits";
+  }
+}
+
+///measurements method if already got the Event 
+MeasurementContainer
+MTDDetLayerMeasurements::measurements( const DetLayer* layer,
+				       const TrajectoryStateOnSurface& startingState,
+				       const Propagator& prop,
+				       const MeasurementEstimator& est) {
+  checkEvent();
+  return measurements(layer, startingState, prop, est, *theEvent);
+}
+
+
+MeasurementContainer
+MTDDetLayerMeasurements::measurements( const DetLayer* layer,
+				       const TrajectoryStateOnSurface& startingState,
+				       const Propagator& prop,
+				       const MeasurementEstimator& est,
+				       const edm::Event& iEvent) {
+  
+  MeasurementContainer result;  
+
+  std::vector<DetWithState> dss = layer->compatibleDets(startingState, prop, est);
+  LogDebug("RecoMTD")<<"compatibleDets: "<<dss.size()<<std::endl;
+  
+  for(std::vector<DetWithState>::const_iterator detWithStateItr = dss.begin();
+      detWithStateItr != dss.end(); ++detWithStateItr){
+
+    MeasurementContainer detMeasurements 
+      = measurements(layer, detWithStateItr->first, 
+                     detWithStateItr->second, est, iEvent);
+    result.insert(result.end(), detMeasurements.begin(), detMeasurements.end());
+  }
+  
+  if (!result.empty()) sort( result.begin(), result.end(), TrajMeasLessEstim());
```

Comment:
could slightly reduce code duplication by putting this reused line into a function like `sortResult(MeasurementContainer& result)` (or even a template function if the result is sometimes a different type)